### PR TITLE
Sort files to provide a fixed order for terraform rendering

### PR DESCRIPTION
**Pull request by manjunjiao**

### CODE REVIEW

**Comment by manjunjiao** [comment]
By introducing this change, the sequence of the files will be in order. 
Terraform plan will be resulting with:
"No changes. Your infrastructure matches the configuration."

**Review by redradrat — CHANGES_REQUESTED** [review]
Let's not introduce a new helper struct and slice. 

Can't we drop all the changes except adding the sort.Slice call, but instead of your proposal we do:

```
sort.Slice(out, func(i, j int) bool {
	return bundle[i].path < bundle[j].path
})
```

That way we'd just sort the already existing out slice, and we know that bundle and out are in the same order.

**Comment by manjunjiao** [comment]
> Let's not introduce a new helper struct and slice.
> 
> Can't we drop all the changes except adding the sort.Slice call, but instead of your proposal we do:
> 
> ```
> sort.Slice(out, func(i, j int) bool {
> 	return bundle[i].path < bundle[j].path
> })
> ```
> 
> That way we'd just sort the already existing out slice, and we know that bundle and out are in the same order.

Thanks for reviewing.

File: pkg/concepts/targets.go [hunk 1]
```diff
@@ -102,13 +104,23 @@ func renderJsonnetConcept(path string, avs *RenderValues, single bool) ([]File,
...
+		return bundle[i].path < bundle[j].path
+	})
+
+	for _, b := range bundle {
+		var m manifest.Manifest
+		if err := yaml.Unmarshal(b.content, &m); err != nil {
+			return nil, err
+		}
+		out = append(out, m)
+	}
```
Comment (redradrat) [inline pkg/concepts/targets.go]: this whole iteration over bundle is not needed as per comment above.

File: pkg/concepts/targets.go [hunk 2]
```diff
@@ -102,13 +104,23 @@ func renderJsonnetConcept(path string, avs *RenderValues, single bool) ([]File,
...
 	out := make(manifest.List, 0, len(extract))
 	for _, m := range extract {
-		out = append(out, m)
 		bundle = append(bundle, File{
 			path:    fmt.Sprintf("%s_%s_%s.yaml", strings.ReplaceAll(m.APIVersion(), "/", "-"), m.Kind(), m.Metadata().Name()),
 			content: []byte(m.String()),
 		})
 	}
 
+	sort.Slice(bundle, func(i, j int) bool {
```
Comment (redradrat) [inline pkg/concepts/targets.go]: we don't want to sort the `bundle` but the `out` list.

We can directly `sort.Slice(out, ... )` just the comparison function passed will compare the entries from bundle.

**Review by redradrat — CHANGES_REQUESTED** [review]

File: pkg/concepts/targets.go (hunk 2 above)
Comment (redradrat) [reply pkg/concepts/targets.go]: please check my snippet here https://github.com/redradrat/kable/pull/24#pullrequestreview-799131358

Comment (manjunjiao) [reply pkg/concepts/targets.go]: Thanks for reviewing. Just one sec. Will resubmit the change.

Comment (manjunjiao) [reply pkg/concepts/targets.go]: > please check my snippet here [#24 (review)](https://github.com/redradrat/kable/pull/24#pullrequestreview-799131358)

Clear. Thanks. Pls see the latest changes.

Comment (redradrat) [reply pkg/concepts/targets.go]: lgtm

that should produce the same functionality. Could you verify it?

**Comment by redradrat** [comment]
@mjiao 
lgtm

that should produce the same functionality. Could you verify it?

**Comment by manjunjiao** [comment]
> @mjiao lgtm
> 
> that should produce the same functionality. Could you verify it?

Yes, it's just tested. Works correctly.

**Comment by redradrat** [comment]
@mjiao thx for the contribution

**Comment by manjunjiao** [comment]
> @mjiao lgtm
> 
> that should produce the same functionality. Could you verify it?

I would also like to mention that the kable terraform provider works great with https://registry.terraform.io/providers/gavinbunney/kubectl/latest/docs/data-sources/kubectl_path_documents

When terraform is upgraded to 0.15, they work together quite well.

**Pull request merged**